File: apps/azimuth/src/routes/pulsify/projects.ts
import {
  createAuthDb,
  createPulsifyDb,
  issues,
  pluginInstallations,
  projects,
  quotas,
  serverMetadata,
} from '@bx-team/stratus/d1';
import { and, count, eq, inArray, ne } from 'drizzle-orm';
import { Hono } from 'hono';
import { z } from 'zod';
import type { Env } from '../../env';
import { aeQuery, sqlStringList } from '../../lib/analytics-sql';
import { ownedProject, resolveOwners, userExists } from '../../lib/pulsify';
import type { SessionVariables } from '../../middleware/auth';

type PulsifyEnv = { Bindings: Env; Variables: SessionVariables };

const createBodySchema = z.object({
  name: z.string().min(1).max(64),
  slug: z
    .string()
    .min(1)
    .max(64)
    .regex(/^[a-z0-9-]+$/),
  type: z.enum(['server', 'plugin', 'mod']),
  description: z.string().max(256).optional(),
});

const verifyBodySchema = z.object({ verified: z.boolean() });

export const projectRoutes = new Hono<PulsifyEnv>();

/**
 * Open (unsuppressed) error fingerprint counts per project — the AE `errors` dataset
 * grouped by fingerprint, minus fingerprints suppressed in the D1 issue registry.
 */
async function countOpenErrors(
  env: Env,
  db: ReturnType<typeof createPulsifyDb>,
  projectIds: string[],
): Promise<Map<string, number>> {
  const [suppressedRows, fingerprintRows] = await Promise.all([
    db
      .select({ projectId: issues.projectId, fingerprint: issues.fingerprint })
      .from(issues)
      .where(and(inArray(issues.projectId, projectIds), ne(issues.status, 'open'))),
    aeQuery<{ project_id: string; fingerprint: string }>(
      env,
      `SELECT index1 AS project_id, blob3 AS fingerprint
       FROM errors
       WHERE index1 IN (${sqlStringList(projectIds)}) AND blob3 != ''
       GROUP BY project_id, fingerprint`,
    ).catch(() => []),
  ]);

  const suppressedByProject = new Map<string, Set<string>>();
  for (const row of suppressedRows) {
    if (!suppressedByProject.has(row.projectId)) suppressedByProject.set(row.projectId, new Set());
    suppressedByProject.get(row.projectId)?.add(row.fingerprint);
  }

  const openByProject = new Map<string, number>();
  for (const row of fingerprintRows) {
    if (suppressedByProject.get(row.project_id)?.has(row.fingerprint)) continue;
    openByProject.set(row.project_id, (openByProject.get(row.project_id) ?? 0) + 1);
  }
  return openByProject;
}

projectRoutes.get('/projects', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  const db = createPulsifyDb(c.env.PULSIFY_DB);

  // Admin view of another user's projects — authorization is a role check on the same endpoint.
  const ownerParam = c.req.query('owner');
  if (ownerParam) {
    if (session.user.role !== 'admin') return c.json({ message: 'Forbidden' }, 403);

    const [rows, owners] = await Promise.all([
      db
        .select({
          id: projects.id,
          name: projects.name,
          slug: projects.slug,
          type: projects.type,
          description: projects.description,
          verified: projects.verified,
          createdAt: projects.createdAt,
        })
        .from(projects)
        .where(eq(projects.ownerId, ownerParam)),
      resolveOwners(createAuthDb(c.env.AUTH_DB), [ownerParam]),
    ]);
    return c.json(rows.map(row => ({ ...row, owner: owners.get(ownerParam) ?? null })));
  }

  const owned = await db
    .select({
      id: projects.id,
      name: projects.name,
      slug: projects.slug,
      type: projects.type,
      description: projects.description,
      createdAt: projects.createdAt,
    })
    .from(projects)
    .where(eq(projects.ownerId, session.user.id));

  if (!owned.length) return c.json([]);

  const projectIds = owned.map(p => p.id);
  const [lastSeenRows, errorsByProject] = await Promise.all([
    db
      .select({ projectId: serverMetadata.projectId, lastSeenAt: serverMetadata.lastSeenAt })
      .from(serverMetadata)
      .where(inArray(serverMetadata.projectId, projectIds)),
    countOpenErrors(c.env, db, projectIds),
  ]);

  const lastSeenByProject = new Map(lastSeenRows.map(r => [r.projectId, r.lastSeenAt]));

  return c.json(
    owned.map(p => ({
      ...p,
      lastSeenAt: lastSeenByProject.get(p.id) ?? null,
      errors: errorsByProject.get(p.id) ?? 0,
    })),
  );
});

projectRoutes.post('/projects', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  const db = createPulsifyDb(c.env.PULSIFY_DB);

  const parsed = createBodySchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) return c.json({ message: parsed.error.message }, 400);
  const { name, slug, type, description } = parsed.data;

  // Confirm the user still exists before writing an owner reference — a valid session
  // can outlive a deleted user, and there is no cross-database FK to catch it.
  if (!(await userExists(createAuthDb(c.env.AUTH_DB), session.user.id))) {
    return c.json({ message: 'User account no longer exists' }, 403);
  }

  // Enforce the per-user project quota (defaults apply when the user has no quota row yet).
  const [[quota], [projectCount]] = await Promise.all([
    db.select({ maxProjects: quotas.maxProjects }).from(quotas).where(eq(quotas.userId, session.user.id)),
    db.select({ count: count() }).from(projects).where(eq(projects.ownerId, session.user.id)),
  ]);
  const maxProjects = quota?.maxProjects ?? 10;
  if ((projectCount?.count ?? 0) >= maxProjects) {
    return c.json({ message: `Project limit reached (${maxProjects}).` }, 403);
  }

  // Plugin/mod names are globally unique — they are the key cross-server error aggregation
  // matches on, so two projects can't claim the same name.
  if (type === 'plugin' || type === 'mod') {
    const [existing] = await db
      .select({ id: projects.id })
      .from(projects)
      .where(and(eq(projects.name, name), inArray(projects.type, ['plugin', 'mod'])));
    if (existing) {
      return c.json({ message: `A plugin or mod named "${name}" already exists.` }, 409);
    }
  }

  const [project] = await db
    .insert(projects)
    .values({ name, slug, type, description, ownerId: session.user.id })
    .returning({
      id: projects.id,
      name: projects.name,
      slug: projects.slug,
      type: projects.type,
    });

  return c.json(project, 201);
});

projectRoutes.delete('/projects/:id', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  const db = createPulsifyDb(c.env.PULSIFY_DB);
  const id = c.req.param('id');

  const [deleted] = await db
    .delete(projects)
    .where(and(eq(projects.id, id), eq(projects.ownerId, session.user.id)))
    .returning({ id: projects.id });

  if (!deleted) return c.json({ message: 'Project not found' }, 404);
  return c.json({ success: true });
});

projectRoutes.patch('/projects/:id/verify', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  if (session.user.role !== 'admin') return c.json({ message: 'Forbidden' }, 403);
  const db = createPulsifyDb(c.env.PULSIFY_DB);
  const id = c.req.param('id');

  const parsed = verifyBodySchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) return c.json({ message: parsed.error.message }, 400);

  const [project] = await db.select({ id: projects.id, type: projects.type }).from(projects).where(eq(projects.id, id));
  if (!project) return c.json({ message: 'Project not found' }, 404);
  if (project.type === 'server') return c.json({ message: 'Only plugin/mod projects can be verified' }, 400);

  const [updated] = await db
    .update(projects)
    .set({ verified: parsed.data.verified, updatedAt: new Date() })
    .where(eq(projects.id, id))
    .returning({ id: projects.id, verified: projects.verified });

  return c.json(updated);
});

projectRoutes.get('/projects/:id/plugins', async c => {
  const session = c.get('authSession');
  if (!session) return c.json({ message: 'Unauthorized' }, 401);
  const db = createPulsifyDb(c.env.PULSIFY_DB);
  const id = c.req.param('id');

  const project = await ownedProject(db, id, session.user.id);
  if (!project) return c.json({ message: 'Project not found' }, 404);
  if (project.type === 'server') return c.json({ message: 'Not applicable for server projects' }, 400);

  const installs = await db
    .select({
      version: pluginInstallations.version,
      enabled: pluginInstallations.enabled,
      lastSeenAt: pluginInstallations.lastSeenAt,
    })
    .from(pluginInstallations)
    .where(eq(pluginInstallations.pluginId, id));

  const total = installs.length;
  const enabled = installs.filter(i => i.enabled).length;

  const versionCounts = new Map<string, number>();
  for (const install of installs) {
    versionCounts.set(install.version, (versionCounts.get(install.version) ?? 0) + 1);
  }

  const versions = [...versionCounts.entries()]
    .sort((a, b) => b[1] - a[1])
    .map(([version, versionCount]) => ({
      version,
      count: versionCount,
      pct: total > 0 ? Math.round((versionCount / total) * 1000) / 10 : 0,
    }));

  const latestVersion = versions[0]?.version ?? null;

  return c.json({
    total_installs: total,
    enabled_installs: enabled,
    latest_version: latestVersion,
    latest_version_adoption: latestVersion ? (versions.find(v => v.version === latestVersion)?.pct ?? 0) : 0,
    versions,
  });
});

File: apps/azimuth/src/routes/pulsify/projects.rs
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::{Json, response::IntoResponse};
use database::models::pulsify as db;
use serde::Deserialize;
use util::{ApiError, ApiResult};
use uuid::Uuid;

use crate::auth::project::OwnedProject;
use crate::auth::session::Session;
use crate::models::pulsify as api;
use crate::state::AppState;

const MAX_SLUG: usize = 64;
const MAX_DESCRIPTION: usize = 256;

#[derive(Debug, Deserialize)]
pub struct OwnerQuery {
    pub owner: Option<Uuid>,
}

#[utoipa::path(get, path = "/pulsify/projects", tag = "pulsify",
    responses((status = 200, body = Vec<api::Project>)))]
pub async fn list(
    State(state): State<AppState>,
    session: Session,
    Query(query): Query<OwnerQuery>,
) -> ApiResult<Json<Vec<api::Project>>> {
    let owner = match query.owner {
        Some(owner) if owner != session.user.id => {
            if !session.user.is_admin() {
                return Err(ApiError::Forbidden("Forbidden".into()));
            }
            owner
        }
        _ => session.user.id,
    };

    let projects = db::projects_of(&state.db, owner).await?;
    let ids: Vec<Uuid> = projects.iter().map(|project| project.id).collect();

    let metadata = db::server_metadata_for(&state.db, &ids).await?;
    let suppressed = db::suppressed_fingerprints(&state.db, &ids).await?;
    let errors = state.analytics.open_error_counts(&ids, &suppressed).await?;

    Ok(Json(
        projects
            .into_iter()
            .map(|project| {
                let last_seen = metadata
                    .iter()
                    .find(|row| row.project_id == project.id)
                    .map(|row| api::iso(row.last_seen_at));

                api::Project {
                    errors: errors.get(&project.id).copied().unwrap_or(0),
                    id: project.id,
                    name: project.name,
                    slug: project.slug,
                    kind: project.kind,
                    description: project.description,
                    verified: project.verified,
                    created_at: api::iso(project.created_at),
                    last_seen_at: last_seen,
                }
            })
            .collect(),
    ))
}

#[utoipa::path(post, path = "/pulsify/projects", tag = "pulsify",
    request_body = api::CreateProject,
    responses((status = 201, body = api::Project), (status = 403), (status = 409)))]
pub async fn create(
    State(state): State<AppState>,
    session: Session,
    Json(body): Json<api::CreateProject>,
) -> ApiResult<(StatusCode, Json<api::Project>)> {
    let name = body.name.trim();
    let slug = body.slug.trim();

    if name.is_empty() || name.chars().count() > MAX_SLUG {
        return Err(ApiError::BadRequest("name must be 1..64 characters".into()));
    }
    if slug.is_empty()
        || slug.len() > MAX_SLUG
        || !slug
            .bytes()
            .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'-')
    {
        return Err(ApiError::BadRequest(
            "slug must match /^[a-z0-9-]+$/ and be 1..64 characters".into(),
        ));
    }
    if !matches!(body.kind.as_str(), "server" | "plugin" | "mod") {
        return Err(ApiError::BadRequest(
            "type must be server, plugin or mod".into(),
        ));
    }
    if body
        .description
        .as_ref()
        .is_some_and(|value| value.chars().count() > MAX_DESCRIPTION)
    {
        return Err(ApiError::BadRequest(
            "description must be at most 256 characters".into(),
        ));
    }

    let quota = db::quota(&state.db, session.user.id).await?;
    if db::count_projects(&state.db, session.user.id).await? >= i64::from(quota.max_projects) {
        return Err(ApiError::Forbidden(format!(
            "Project limit of {} reached",
            quota.max_projects
        )));
    }

    let project = db::create_project(
        &state.db,
        session.user.id,
        name,
        slug,
        &body.kind,
        body.description.as_deref().map(str::trim),
    )
    .await
    .map_err(|error| match error {
        // Both the slug and the global plugin name are unique; either is a conflict, not a 500.
        database::Error::Database(ref inner) if inner.is_unique_violation() => {
            ApiError::Conflict("A project with this name or slug already exists".into())
        }
        other => ApiError::Database(other),
    })?;

    Ok((
        StatusCode::CREATED,
        Json(api::Project {
            id: project.id,
            name: project.name,
            slug: project.slug,
            kind: project.kind,
            description: project.description,
            verified: project.verified,
            created_at: api::iso(project.created_at),
            last_seen_at: None,
            errors: 0,
        }),
    ))
}

#[utoipa::path(delete, path = "/pulsify/projects/{id}", tag = "pulsify",
    params(("id" = Uuid, Path, description = "Project id")),
    responses((status = 204), (status = 404)))]
pub async fn delete(
    State(state): State<AppState>,
    owned: OwnedProject,
) -> ApiResult<impl IntoResponse> {
    db::delete_project(&state.db, owned.project.id, owned.user.id).await?;
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Debug, serde::Serialize, utoipa::ToSchema)]
pub struct Verified {
    pub id: Uuid,
    pub verified: bool,
}

/// Verification is what unlocks cross-server aggregation, so only an admin may grant it.
#[utoipa::path(patch, path = "/pulsify/projects/{id}/verify", tag = "pulsify",
    params(("id" = Uuid, Path, description = "Project id")),
    request_body = api::SetVerified,
    responses((status = 200, body = Verified), (status = 403), (status = 404)))]
pub async fn verify(
    State(state): State<AppState>,
    _admin: crate::auth::session::AdminSession,
    Path(id): Path<Uuid>,
    Json(body): Json<api::SetVerified>,
) -> ApiResult<Json<Verified>> {
    if !db::set_verified(&state.db, id, body.verified).await? {
        return Err(ApiError::NotFound("Project not found".into()));
    }

    Ok(Json(Verified {
        id,
        verified: body.verified,
    }))
}

#[utoipa::path(get, path = "/pulsify/projects/{id}/plugins", tag = "pulsify",
    params(("id" = Uuid, Path, description = "Project id")),
    responses((status = 200, body = api::PluginInsights), (status = 400), (status = 404)))]
pub async fn plugins(
    State(state): State<AppState>,
    owned: OwnedProject,
) -> ApiResult<Json<api::PluginInsights>> {
    owned.require_installable()?;

    let installations = db::installations_of(&state.db, owned.project.id).await?;
    let total = installations.len() as u64;
    let enabled = installations.iter().filter(|row| row.enabled).count() as u64;

    let mut by_version: std::collections::HashMap<String, u64> = std::collections::HashMap::new();
    for installation in &installations {
        *by_version.entry(installation.version.clone()).or_default() += 1;
    }

    let mut versions: Vec<api::VersionShare> = by_version
        .into_iter()
        .map(|(version, count)| api::VersionShare {
            version,
            count,
            pct: api::percent(count, total),
        })
        .collect();
    versions.sort_by(|a, b| {
        b.count
            .cmp(&a.count)
            .then_with(|| a.version.cmp(&b.version))
    });

    // The most installed version, not the newest one: adoption is what this page is about.
    let latest = versions.first();

    Ok(Json(api::PluginInsights {
        total_installs: total,
        enabled_installs: enabled,
        latest_version: latest.map(|share| share.version.clone()),
        latest_version_adoption: latest.map_or(0.0, |share| share.pct),
        versions,
    }))
}

#[utoipa::path(get, path = "/pulsify/projects/{id}/installations", tag = "pulsify",
    params(("id" = Uuid, Path, description = "Project id")),
    responses((status = 200, body = api::Installations), (status = 400), (status = 404)))]
pub async fn installations(
    State(state): State<AppState>,
    owned: OwnedProject,
) -> ApiResult<Json<api::Installations>> {
    owned.require_server()?;

    let rows = db::installations_on(&state.db, owned.project.id).await?;

    Ok(Json(api::Installations {
        installations: rows
            .into_iter()
            .map(|row| api::Installation {
                plugin_id: row.plugin_id,
                name: row.name,
                slug: row.slug,
                version: row.version,
                enabled: row.enabled,
                share_errors: row.share_errors,
                last_seen_at: api::iso(row.last_seen_at),
            })
            .collect(),
    }))
}

/// Lets a server owner stop sharing one plugin's errors with its author.
#[utoipa::path(patch, path = "/pulsify/projects/{id}/installations/{plugin_id}", tag = "pulsify",
    params(
        ("id" = Uuid, Path, description = "Server project id"),
        ("plugin_id" = Uuid, Path, description = "Installed plugin project id")
    ),
    request_body = api::SetShareErrors,
    responses((status = 204), (status = 404)))]
pub async fn set_share_errors(
    State(state): State<AppState>,
    owned: OwnedProject,
    Path(path): Path<(Uuid, Uuid)>,
    Json(body): Json<api::SetShareErrors>,
) -> ApiResult<impl IntoResponse> {
    let (_, plugin_id) = path;
    owned.require_server()?;

    if !db::set_share_errors(&state.db, owned.project.id, plugin_id, body.share_errors).await? {
        return Err(ApiError::NotFound("Installation not found".into()));
    }

    Ok(StatusCode::NO_CONTENT)
}
